File: src/main/java/ludwig/model/LiteralNode.java
package ludwig.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import ludwig.utils.NodeUtils;

public class LiteralNode extends Node<LiteralNode> {
    private String text;
    @JsonIgnore
    private Object value;

    public LiteralNode() {
    }

    public LiteralNode(String text) {
        this(text, NodeUtils.parseLiteral(text));
    }

    private LiteralNode(String text, Object value) {
        this.text = text;
        this.value = value;
    }

    public String text() {
        return text;
    }

    public Object value() {
        if (value == null) {
            value = NodeUtils.parseLiteral(text);
        }
        return value;
    }

    @Override
    public <T> T accept(NodeVisitor<T> visitor) {
        return visitor.visitLiteral(this);
    }

    public static LiteralNode ofValue(Object value) {
        return new LiteralNode(NodeUtils.formatLiteral(value), value);
    }

    @Override
    public String toString() {
        return text;
    }
}

File: src/main/java/ludwig/model/LiteralNode.kt
package ludwig.model

import com.fasterxml.jackson.annotation.JsonIgnore
import ludwig.utils.NodeUtils

class LiteralNode(val text: String) : Node<LiteralNode>() {

    //@JsonIgnore
    val value: Any? by lazy {
        NodeUtils.parseLiteral(text)
    }

    override fun <T> accept(visitor: NodeVisitor<T>): T {
        return visitor.visitLiteral(this)
    }

    override fun toString(): String {
        return text
    }

    companion object {

        fun ofValue(value: Any): LiteralNode {
            return LiteralNode(value.toString())
        }
    }
}
